Zoltan/defining default allocation

zoltanszocs opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and
review the checklist below.

Contributors guide: ./CONTRIBUTING.md
-->

## Checklist
<!--
Remove items that do not apply. For completed items, change [ ] to [x] or you can click the checkboxes once your 
pull-request is published.
-->

- [ ] Keep pull requests small so they can be easily reviewed.
- [ ] Update the documentation, please follow the [numpy style guide](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] Update tests.
- [ ] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `feature`, `ui`, `change`, `documentation`, `breaking`, `ci`
      as they show up in the changelog.
- [ ] Link this PR to related issues by using [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
- [ ] Add a milestone to the PR for the intended release.
- [ ] Request a review from another developer.

<!--
NOTE: these things are not required to open a PR and can be done afterwards,
while the PR is open.
-->

ughstudios committed:
[CODE] Fixes crash when app tries to serialize pathlib Paths

Resolves #22

- The issue here is that bd.projects.get_base_directories() will return a Path object which cannot be serialized. We can address this by simply converting it to a string before saving it out.

===== TESTS ======
I tested by ensuring my local copy of AB would boot up without issues. Prior to this, I was getting errors about the json file not being readable or writable with json.load or json.dump

cmutel committed:
Merge pull request #23 from cauldron/22-app-will-crash-if-settings-is-invalid

app will crash if settings is invalid

zoltanszocs committed:
Fix comment

zoltanszocs committed:
Add editing with combobox for "Default Allocation"

zoltanszocs committed:
Fix multiple model-view issues

- sorting after changing a  def alloc field does not always work automatically (sorting column and order is reset by the update_proxy_model() call)
- checking the read-only checkboxes sometimes triggers a reorder of the table (probably because of the sync() call)
- application crashes when flushing the databases (even without changes),
  because flush() triggers a sync(), which does not call begin / end ModelReset()

zoltanszocs committed:
Correct initial sort

zoltanszocs committed:
Prevent double click from opening the database in some cases

- on the read-only flag: here we expect clicks to toggle the flag
- on the default allocation column, when the database is not read-only: in this case we expect the item editor for the cell to open

zoltanszocs committed:
Fix usage of createIndex

createIndex(row, col) only works with QTableView and with no proxy model, in all these cases index is the correct method to call.

zoltanszocs committed:
Add combobox for showing / editing default allocation for processes

zoltanszocs committed:
Minor fix

zoltanszocs committed:
Minor refactor for testing

zoltanszocs committed:
Add test for property editor

zoltanszocs committed:
Add handling for non-numerical properties

zoltanszocs committed:
Add custom allocation editor

zoltanszocs committed:
Open the custom allocation dialog faster when the user selects "Custom..."

zoltanszocs committed:
Add sorting and keep selected line visible

zoltanszocs committed:
Fix plain text edit too big

zoltanszocs committed:
Add comments

cmutel commented:
@zoltanszocs This is good progress! The double-click to change database default allocation is definitely not intuitive, but I don't think we need to worry about that right now.

I updated https://github.com/cauldron/activity-browser/issues/33 with more detail on the remaining tasks.

> Database processes which have the type multifunctional cannot be dragged into calculation setups

This is blocked by https://github.com/brightway-lca/multifunctional/issues/23

>  Uses should be able to see if the given allocation property key is present and non-zero for each functional exchange of each multifunctional process in a database.

I don't see the allocation values, not sure where the bug is. According to [the tests](https://github.com/brightway-lca/multifunctional/blob/main/tests/test_store_allocation_data.py) this information is being stored correctly?

zoltanszocs committed:
Fix def alloc column for not multifunctional db

For not multifunctional db:
- default allocatin can not be edited now
- its value is "Not applicable"

cmutel committed:
Factorize for better speed on umfpack

cmutel committed:
Bump dependencies

cmutel committed:
Convert inputs to ints

cmutel committed:
Merge branch 'main' into pr/45

zoltanszocs committed:
Redo allocations on allocation method change

zoltanszocs committed:
Merge branch 'brightway25' of https://github.com/zoltanszocs/activity-browser into zoltan/defining-default-allocation

zoltanszocs committed:
Fix merge errors

zoltanszocs committed:
Fix another merge error

zoltanszocs committed:
Add missing set_read_only call

zoltanszocs committed:
Update allocations when properties or production amount change

zoltanszocs committed:
Only show statusbar message for a limited time

zoltanszocs committed:
Make the allocation factor cell read-only

zoltanszocs committed:
Redo allocation only for multifunctional databases

zoltanszocs committed:
Disconnect the update signal when changing the values in the combobox

Each line of the changes would trigger an allocation recalculation.

zoltanszocs committed:
Disable the save button when nothing is selected

zoltanszocs committed:
Fix random column layout in inventory

Use a fixed layout and change some columns (order, name add `Type`)

zoltanszocs committed:
Add back key column, it is currently used

zoltanszocs committed:
Fix getting the selected item

zoltanszocs committed:
Fix the case of no default allocation + cancel on the editor

zoltanszocs committed:
Undo changes to commontasks

Rename columns with different approach and remove key column

zoltanszocs committed:
Fix the column reorganize for USEEIO and ecoinvent databases

In these databases, some columns are missing, so the previous change with fixed column names gave errors.

zoltanszocs committed:
ComboBoxDelegate can display now colored rows

zoltanszocs committed:
Execute property checks for the whole database or a process only

zoltanszocs committed:
Add colors to the database custom allocation combobox

zoltanszocs committed:
Simplify inventory model, fix USEEIO filter exception

The reason for the exception was, that the self.filterable_columns contained the already removed ISIC column

zoltanszocs committed:
Fix old exception on reset search

zoltanszocs committed:
Fix old bug: ISIC column appears when searching in USEEIO

zoltanszocs committed:
Fix old bug: clear search with empty edit box cleared the filters also

cmutel committed:
Explicit parameter imports

cmutel committed:
Merge branch 'zoltan/defining-default-allocation' of https://github.com/zoltanszocs/activity-browser into pr/45

zoltanszocs committed:
Fix search finding empty cells when searching for 'nan'

zoltanszocs committed:
Do not allow editing of read-only processes

zoltanszocs committed:
Add test version of the button editor for custom allocation

zoltanszocs committed:
Add hyperlink variant

zoltanszocs committed:
Fix column handling for empty database

zoltanszocs committed:
Add colors to activity combobox

zoltanszocs committed:
Fix column count for empty database

zoltanszocs committed:
Allow the user to select a property by double click

zoltanszocs committed:
Do not save the databases if the allocation did not change

zoltanszocs committed:
Select button should be disabled at start, enabled when something is selected

zoltanszocs committed:
Move status text setting to after the table filling, it uses rowCount()

zoltanszocs committed:
Add tests for the property editor

zoltanszocs committed:
Fix the testcase blocking test execution

zoltanszocs committed:
Fix column names lost in refactor

zoltanszocs committed:
Fix sort exception in USEEIO due to nan values

cmutel committed:
Merge branch 'brightway25' into pr/45

cmutel committed:
Update some text labels

cmutel committed:
Don't process when only need to reallocate

cmutel committed:
Change activity tab labels

cmutel committed:
Use hyperlink instead of button for allocation column

cmutel committed:
Update dependencies

cmutel committed:
Merge branch 'brightway25' into pr/45

zoltanszocs committed:
Update dependencies according to recipe-25/meta.yaml

zoltanszocs committed:
Fix exception

zoltanszocs committed:
Fix column renaming

zoltanszocs committed:
Remove combobox and pushbutton editing of custom allocation

zoltanszocs committed:
Update custom allocation colors to colorblind friendlier colors

zoltanszocs committed:
Show `equal` allocation always as green.

zoltanszocs committed:
Sort combobox values alphabetically
